### PR TITLE
term agg

### PR DESCRIPTION
**Pull request by PSeitz**
- support order property on term aggregations
- return error on missing sub aggregation
- fix is_fast for bytes field
- minor fixes

### CODE REVIEW

File: src/aggregation/agg_result.rs [hunk 1]
```diff
@@ -18,55 +17,76 @@ use super::intermediate_agg_result::{
...
-
+    ) -> crate::Result<Self> {
         // Important assumption:
         // When the tree contains buckets/metric, we expect it to have all buckets/metrics from the
         // request
-        if let Some(buckets) = results.buckets {
-            result.extend(buckets.into_iter().zip(req.buckets.values()).map(
-                |((key, bucket), req)| {
-                    (
+        let mut result: HashMap<_, _> = if let Some(buckets) = results.buckets {
```
Comment (fulmicoton) [inline src/aggregation/agg_result.rs]: This is very hard to follow.
Can you:
- isolate the two bodies in independent functions?
- add an assert on the stuff being zipped being of the same size.
- use an iterative style maybe?

File: src/aggregation/bucket/mod.rs [hunk 2]
```diff
@@ -11,8 +11,130 @@ mod histogram;
...
+        order: Order::Desc,
+    };
+
+    let order_str = serde_json::to_string(&order).unwrap();
+    assert_eq!(order_str, "{\"_key\":\"desc\"}");
+    let order_deser = serde_json::from_str(&order_str).unwrap();
+
+    assert_eq!(order, order_deser);
+
+    let order_deser: serde_json::Result<CustomOrder> = serde_json::from_str(&"{}");
```
Comment (fulmicoton) [inline src/aggregation/bucket/mod.rs]:
```suggestion
    let order_deser: serde_json::Result<CustomOrder> = serde_json::from_str("{}");
```

File: src/aggregation/intermediate_agg_result.rs [hunk 3]
```diff
@@ -218,7 +221,7 @@ impl IntermediateBucketResult {
                     })
                     .collect();
 
-                std::mem::swap(entries_left, &mut buckets);
+                std::mem::swap(buckets_left, &mut buckets);
```
Comment (fulmicoton) [inline src/aggregation/intermediate_agg_result.rs]: the use of mem::swap is useless here.

File: src/aggregation/intermediate_agg_result.rs [hunk 4]
```diff
@@ -246,22 +255,75 @@ impl IntermediateTermBucketResult {
...
+                })
             })
-            .collect();
-        buckets.sort_unstable_by_key(|bucket| std::cmp::Reverse(bucket.doc_count));
+            .collect::<crate::Result<_>>()?;
+
+        let order = req.order.order;
+        match req.order.target {
+            OrderTarget::Key => {
+                buckets.sort_by(|bucket1, bucket2| {
```
Comment (fulmicoton) [inline src/aggregation/intermediate_agg_result.rs]: nitpick, prefer left and right than 1 / 2

File: src/aggregation/intermediate_agg_result.rs (hunk 3 above)
Comment (fulmicoton) [reply src/aggregation/intermediate_agg_result.rs]: buckets_left = buckets;

File: src/aggregation/intermediate_agg_result.rs [hunk 5]
```diff
@@ -246,22 +255,75 @@ impl IntermediateTermBucketResult {
...
+                    if req.order.order == Order::Desc {
+                        bucket1
```
Comment (fulmicoton) [inline src/aggregation/intermediate_agg_result.rs]: very local, but there is probably a nicer way to write this.

**Review by fulmicoton — APPROVED** [review]
Please address inline comments.

Also can you do a pass on the PR and see if the code can be simplified?

File: src/aggregation/intermediate_agg_result.rs (hunk 3 above)
Comment (PSeitz) [reply src/aggregation/intermediate_agg_result.rs]: entries_left is a &mut binding from a match pattern and buckets is a temporary variable. I tried some things, but couldn't find a solution other than mem::swap

Comment (fulmicoton) [reply src/aggregation/intermediate_agg_result.rs]: `*entries_left = buckets` did not work?

File: src/aggregation/intermediate_agg_result.rs (hunk 5 above)
Comment (PSeitz) [reply src/aggregation/intermediate_agg_result.rs]: I did a minor improvement on it
